firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/Chunks.cpp

```diff
@@ -0,0 +1,176 @@
+      auto final_status = chunks.input(val);
+
+      THEN("the final status should be ok") {
+        REQUIRE(status == PF::Protocols::ChunkInputStatus::ok);
+        REQUIRE(output_status == PF::Protocols::ChunkOutputStatus::waiting);
+        REQUIRE(final_status == PF::Protocols::ChunkInputStatus::invalid_length);
+      }
+    }
+  }
+}
+
+SCENARIO(
+    "Protocols::ChunkSplitter behaves correctly across chunk boundary(102 bytes)", "[chunks]") {
+  GIVEN("A chunk splitter") {
+    constexpr size_t buffer_size = 102;
+    PF::Protocols::ChunkSplitter<buffer_size, char> chunks;
+
+    WHEN("the input data is within bounds") {
+      uint8_t val = 128;
+      PF::Protocols::ChunkInputStatus status;
+      for (size_t i = 0; i < buffer_size; ++i) {
+        status = chunks.input(val);
+      }
+
+      THEN("the final status should be ok") {
+        REQUIRE(status == PF::Protocols::ChunkInputStatus::ok);
+      }
+    }
+
+    WHEN("the input data exceeds bounds without output being called") {
+      uint8_t val = 128;
+      PF::Protocols::ChunkInputStatus status;
+      for (size_t i = 0; i < buffer_size; ++i) {
+        status = chunks.input(val);
+      }
+
+      for (size_t i = 0; i < buffer_size; ++i) {
+        status = chunks.input(val);
+      }
+
+      THEN("the final status should be invalid_length") {
+        REQUIRE(status == PF::Protocols::ChunkInputStatus::invalid_length);
+      }
+    }
+
+    WHEN("the input data exceeds bounds after output is called, without delimiter") {
+      uint8_t val = 128;
+      PF::Protocols::ChunkInputStatus status;
+      for (size_t i = 0; i < buffer_size; ++i) {
+        status = chunks.input(val);
+      }
+      PF::Util::Vector<char, buffer_size> buffer;
+      PF::Protocols::ChunkOutputStatus output_status = chunks.output(buffer);
+      auto final_status = chunks.input(val);
+
+      THEN("the final status should be ok") {
+        REQUIRE(status == PF::Protocols::ChunkInputStatus::ok);
+        REQUIRE(output_status == PF::Protocols::ChunkOutputStatus::waiting);
+        REQUIRE(final_status == PF::Protocols::ChunkInputStatus::invalid_length);
+      }
+    }
+
+    WHEN("output is called on an input of invalid length") {
+      uint8_t val = 128;
+      PF::Protocols::ChunkInputStatus status;
+      for (size_t i = 0; i < buffer_size; ++i) {
+        status = chunks.input(val);
+      }
+      auto final_status = chunks.input(val);
+      PF::Util::Vector<char, buffer_size> buffer;
+      PF::Protocols::ChunkOutputStatus output_status = chunks.output(buffer);
+
+      THEN("the final status should be invalid_length") {
+        REQUIRE(status == PF::Protocols::ChunkInputStatus::ok);
+        REQUIRE(final_status == PF::Protocols::ChunkInputStatus::invalid_length);
+        REQUIRE(output_status == PF::Protocols::ChunkOutputStatus::invalid_length);
+      }
+    }
+
+    WHEN("the input data exceeds bounds after output is called, with delimiter") {
```

Comment:
Is this equivalent to "WHEN the splitter's buffer is filled with input data and the output method is called and more data is input into the splitter"?

Maybe it would be clearer instead to move some of these WHENs into a different GIVEN, e.g. GIVEN a splitter which has been filled with input data:

- WHEN more data is input, THEN an invalid_length input error is returned
- WHEN input is consumed by the output method and more data is input, THEN the input returns ok

and so on.

Reply:
Updated test cases as suggested